sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -940,20 +939,19 @@ object SparkSession extends Logging {
         options.foreach { case (k, v) => session.initialSessionOptions.put(k, v) }
         setDefaultSession(session)
         setActiveSession(session)
-
-        // Register a successfully instantiated context to the singleton. This should be at the
-        // end of the class definition so that the singleton is updated only if there is no
-        // exception in the construction of the instance.
-        sparkContext.addSparkListener(new SparkListener {
-          override def onApplicationEnd(applicationEnd: SparkListenerApplicationEnd): Unit = {
-            defaultSession.set(null)
-          }
-        })
+        registerSessionListenerOnContext(sparkContext)
       }
 
       return session
     }
 
+    private def registerSessionListenerOnContext(sparkContext: SparkContext): Unit = {
```

Comment:
can we move this method closer to where we define `_sessionListener` and `sessionListenerRegistered`?

Reply:
maybe move _sessionListener definition inside of sessionListenerRegistered? _sessionListener seems useless outside of the sessionListenerRegistered.
Also, maybe add some tests?

Reply:
yea a test would be better